Make Decoder implement Send

h33p opened the pull request:
Making all handlers references to arrays seemed like the cleanest solution to me, and without any severe downside. If speaking at micro-optimization levels, initializing the tables will perform size checks in release mode as well. Prefixes were changed into an array as well. Prefixes and handler tables are accessed with array indexing. This could lead to potential overflow checks taking place, but if LLVM is smart enough about it, it should eliminate them on release mode. So I'm unsure of the performance impact, if any, these changes would lead to. I could change it to just implement `Send`, but any input would be welcome.

h33p commented:
Ohhh that's my bad, I will get that fixed

h33p commented:
Okay, I also added unwrap lint ignores, since the input is always supposed to be sized 256, but if you want me to change that, I could add back debug assertion, and use transmute for the conversion.

wtfsck commented on src/rust/iced-x86/src/decoder.rs:
```diff
@@ -354,25 +354,25 @@ pub struct Decoder<'a> {
 	instr_start_data_ptr: usize,
 
 	// These are verified to have exactly 0x100 elements, and they're static, so we don't need fat pointers (slices).
```
This comment can be removed now

wtfsck reviewed (changes requested):
Could you also add a compile check that `Decoder` implements `Send`, see:

https://github.com/icedland/iced/blob/2334aa4a6dd56c4f925339f969bdca2b850d8b6d/src/rust/iced-x86/src/iced_error.rs#L16-L18

wtfsck commented:
> EDIT: Might actually want to change it, since my changes currently depend on `min_const_generics`

It compiles with 1.47.0 or later, I think min_const_generics was a later version?

wtfsck commented:
> [...] So I'm unsure of the performance impact, if any, these changes would lead to. I could change it to just implement Send, but any input would be welcome.

I didn't notice a perf regression, but I'll test again later before merging.

h33p commented:
> > EDIT: Might actually want to change it, since my changes currently depend on `min_const_generics`
> 
> It compiles with 1.47.0 or later, I think min_const_generics was a later version?

That's my bad I'm not making sense of what I'm talking, I thought traits on longer arrays depended on const generics, but it appears that it was added in 1.47.0. Is bumping requirements to the version okay?

Also, I added the compile-time trait check. It appears that Sync is also implemented by this change, so I bound it as well. I tried to do it with _TraitsCheck, but for some reason adding non-existent trait bound, like `Extend<i32>` did not result in compile-time error, so I instead added it on the struct itself, where it seems to function as expected.

wtfsck commented:
> Is bumping requirements to the version okay?

No, I just bumped it to 1.41.0. This version or later is used by many Linux distros.

h33p commented:
Alright, I got rid of `TryFrom` use in those places, instead, directly implementing the conversion. I have verified it compiles on 1.41.0, so all should be good.

wtfsck commented:
I just released 1.11.2 with these changes.
